explorer/syntax/lexer.lpp

```diff
@@ -296,66 +296,79 @@ operand_start         [(A-Za-z0-9_\"]
+#*(\") {
+  // Found a sequence of double quote characters.
   Carbon::StringLexHelper str_lex_helper(yytext, yyscanner, context);
   const std::string& s = str_lex_helper.str();
   const int hashtag_num = s.find_first_of('"');
-  const int leading_quotes = s.size() - hashtag_num;
-  if (leading_quotes == 3 && hashtag_num > 0) {
-    // Check if it's a single-line string, like #"""#.
-    // TODO: Extend with other single-line string cases, like #""""#, based on
-    // the definition of block string in the design doc.
-    if (Carbon::ReadHashTags(str_lex_helper, hashtag_num)) {
-      return Carbon::ProcessSingleLineString(str_lex_helper.str(), context,
-                                             hashtag_num);
-    } else if (str_lex_helper.is_eof()) {
-      return CARBON_SIMPLE_TOKEN(END_OF_FILE);
-    }
-  } else if (!str_lex_helper.Advance()) {
+
+  if (!str_lex_helper.Advance()) {
     return CARBON_SIMPLE_TOKEN(END_OF_FILE);
   }
-  // 3 quotes indicates multi-line, otherwise it'll be one.
-  const bool multi_line = leading_quotes == 3;
 
   while (!str_lex_helper.is_eof()) {
     switch (str_lex_helper.last_char()) {
       case '\n':  // Fall through.
       case '\v':  // Fall through.
       case '\f':  // Fall through.
       case '\r':
-        if (!multi_line) {
-          return context.RecordSyntaxError(
-              llvm::formatv("missing closing quote in single-line string: {0}",
-                            str_lex_helper.str()));
-        }
-        str_lex_helper.Advance();
+        return context.RecordSyntaxError(
+            llvm::formatv("missing closing quote in single-line string: {0}",
+                          str_lex_helper.str()));
         break;
       case '"':
-        if (multi_line) {
-          // Check for 2 more '"'s on block string.
-          if (!(str_lex_helper.Advance() &&
-                str_lex_helper.last_char() == '"')) {
-            continue;
-          }
-          if (!(str_lex_helper.Advance() &&
-                str_lex_helper.last_char() == '"')) {
+        if (Carbon::ReadHashTags(str_lex_helper, hashtag_num)) {
+          // Reach closing quotes, break out of the loop.
+          return Carbon::ProcessSingleLineString(str_lex_helper.str(), context,
+                                                 hashtag_num);
+        }
+        break;
+      case '\\':
+        if (Carbon::ReadHashTags(str_lex_helper, hashtag_num)) {
+          // Read the escaped char.
+          if (!str_lex_helper.Advance()) {
             continue;
           }
-          // Now we are at the last " of """.
+          // Read the next char.
+          str_lex_helper.Advance();
+        }
+        break;
+      default:
+        str_lex_helper.Advance();
+    }
+  }
+
+  return CARBON_SIMPLE_TOKEN(END_OF_FILE);
+}
+
+#*(\'\'\') {
+  // Raw string literal.
+  // yytext (the token that matches the above regex) and chars scanned by
+  // str_lex_helper hold the source text, not the string the source represents.
```

Comment:
```suggestion
```

This comment might belong more on StringLexHelper, but it applies to both versions of the parser equally. I'd suggest removing it though -- I'm not sure it's necessary given the hashtag_num setting, it feels like that should be clarifying if someone's unsure.

Reply:
Removed the comment